all reduce non strict

avikchaudhuri opened the pull request:
Summary:
Some distributed collectives like `all_reduce` have special handling in Dynamo, where they are mapped to functional collectives. Non-strict was previously blind to such mappings, which means using them would fail to trace. Here we show how intercepting them in non-strict's torch function mode can mimic this remapping logic. More ops to follow.

Side note: a recently added distributed test was in the wrong place, making the expected failures for non-strict not fire because we weren't actually generating those tests to begin with! Now fixed.

Test Plan: moved and updated test

Differential Revision: D69607140


cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

tugsbayasgalan commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2779,6 +2780,20 @@ def all_reduce(tensor, op=ReduceOp.SUM, group=None, async_op=False):
+    # Let's redirect to a torch function mode that can mimic this logic outside Dynamo
+    # (e.g., non-strict export implements such a torch function mode).
+    relevant_args = (tensor,)
+    if has_torch_function(relevant_args):
```
Does this have performance implications? We would be doing this in eager mode path. Is there a way to turn it off in eager and only turn it on in export?

avikchaudhuri replied:
These functions are implemented in C++ and supposed to be super fast, and probably will be dominated by `all_reduce` itself. We'll see if something comes up in tests...

tugsbayasgalan replied:
I think there is some previous art like: https://www.internalfb.com/code/fbsource/[7b844baf6c235ba6431ed1808eea9302c560f5fd]/xplat/caffe2/torch/nn/parameter.py?lines=92

avikchaudhuri replied:
I'm not very comfortable disabling torch function mode, what if there are some that are set up elsewhere to handle inputs or what not...?

bdhirsh replied:
cc @wconstab just in case you have any opinions. Given that most of our torch.* API's have torch_function handlers, adding them for the `c10d_functional` API's doesn't seem like it should be an overhead killer.

angelayi commented on torch/_export/non_strict_utils.py:
```diff
@@ -618,6 +619,19 @@ class _NonStrictTorchFunctionHandler(torch.overrides.TorchFunctionMode):
     """
 
     def _override(self, func, args, kwargs):
+        if func is torch.distributed.all_reduce:
```
do you want to make this more general? for all functional collectives?

avikchaudhuri replied:
yeah, after this lands / doesn't get reverted I'll do others. Just want to feel this out first.

avikchaudhuri committed:
all reduce non strict (#147133)

Summary:

Some distributed collectives like `all_reduce` have special handling in Dynamo, where they are mapped to functional collectives. Non-strict was previously blind to such mappings, which means using them would fail to trace. Here we show how intercepting them in non-strict's torch function mode can mimic this remapping logic. More ops to follow.

Side note: a recently added distributed test was in the wrong place, making the expected failures for non-strict not fire because we weren't actually generating those tests to begin with! Now fixed.

Test Plan: moved and updated test

Reviewed By: tugsbayasgalan

Differential Revision: D69607140
